src/lib/db.ts

```diff
@@ -367,23 +369,36 @@ export class DB {
   }
 
   private setConnectionHandlers(userRef: firebase.database.Reference) {
+    if (this.groupOnDisconnect) {
+      this.groupOnDisconnect.cancel();
+    }
+    const groupDisconnectedRef = userRef.child("disconnectedTimestamp");
+    this.groupOnDisconnect = groupDisconnectedRef.onDisconnect();
+    this.groupOnDisconnect.set(firebase.database.ServerValue.TIMESTAMP);
+
     if (this.connectedRef) {
       this.connectedRef.off("value");
     }
     else if (!this.connectedRef) {
       this.connectedRef = firebase.database().ref(".info/connected");
     }
-    this.connectedRef.on("value", (snapshot) => {
-      if (snapshot && snapshot.val()) {
-        userRef.child("connectedTimestamp").set(firebase.database.ServerValue.TIMESTAMP);
-      }
-    });
+    // once() ensures that the connected timestamp is set before resolving
```

Comment:
This pattern is verbose, but I think this is the only way to guarantee the timestamp is set before resolving.